cpp/src/arrow/flight/sql/client.cc

```diff
@@ -802,6 +802,157 @@ ::arrow::Result<CancelResult> FlightSqlClient::CancelQuery(
   return Status::IOError("Server returned unknown result ", result.result());
 }
 
+::arrow::Result<std::vector<SetSessionOptionResult>> FlightSqlClient::SetSessionOptions(
+    const FlightCallOptions& options,
+    const std::vector<SessionOption>& session_options) {
+  flight_sql_pb::ActionSetSessionOptionsRequest request;
+  for (const SessionOption& in_opt : session_options) {
+    flight_sql_pb::SessionOption* opt = request.add_session_options();
+    const std::string& name = in_opt.option_name;
+    opt->set_option_name(name);
+
+    const SessionOptionValue& value = in_opt.option_value;
+    if (value.index() == std::variant_npos)
+      return Status::Invalid("Undefined SessionOptionValue type ");
+    switch (static_cast<SessionOptionValueType>(value.index())) {
```

Comment:
I think the way you're supposed to do this is `std::visit`, and then we don't need the extra enum.

Reply:
Probably worth doing to avoid the code duplication we'd discussed previously—on it.